EIPS/eip-4906.md

```diff
@@ -0,0 +1,83 @@
+---
+eip: 4906
+title: ERC-721 Metadata Update Extension
+description: Add a `MetadataUpdate` event to ERC-721.
+author: Anders (@0xanders), Lance (@LanceSnow)
+discussions-to: https://ethereum-magicians.org/t/eip4906-erc-721-erc-1155-metadata-update-extension/8588
+status: Draft
+type: Standards Track
+category: ERC
+created: 2022-03-13
+requires: 165, 721
+---
+
+## Abstract
+
+This standard is an extension of [ERC-721](./eip-721.md). It proposes adding a `MetadataUpdate` event to ERC-721 tokens.
+
+## Motivation
+
+Many [ERC-721](./eip-721.md) contracts emit a custom event when a token's metadata is changed. It is easy to update metadata of one NFT by specific event, but it is difficult for third-party platforms such as NFT marketplace to update metadata of many NFTs based on custom events.
+
+Having a standard `MetadataUpdate` event will make it easy for third-party platforms to timely update metadata of many NFTs.
+
+## Specification
+
+The keywords “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.
+
+The **metadata update extension** is OPTIONAL for ERC-721/ERC-1155 contracts.
+
+
+```solidity
+/// @title ERC-721 Metadata Update Extension
+interface IERC4906 is IERC165{
+    /// @dev This event emits when the metadata of a token is changed.
+    /// So that the third-party platforms such as NFT market could
+    /// timely update the images and related attributes of the NFT
+    event MetadataUpdate(uint256 _tokenId);
+}
+```
+
+The `MetadataUpdate` event MUST be emitted when the metadata of a token is changed.
```

Comment:
I'd like to see the EIP explicitly state exactly when the event should/shouldn't be emitted. For example, these cases:

 - When a token is minted after the contract is deployed?
 - When a token is minted in the constructor?
 - When a token is burned?
 - When the `tokenURI` changes, but the JSON does not.<- I would assume don't emit?
 - When the `tokenURI` changes, and the JSON also changes. <- I would assume emit
 - When the `tokenURI` does not change, but the JSON does. <- I would also assume emit

Reply:
> * When a token is minted after the contract is deployed?
> * When a token is minted in the constructor?
> * When a token is burned?
> * When the `tokenURI` changes, but the JSON does not.<- I would assume don't emit?
> * When the `tokenURI` changes, and the JSON also changes. <- I would assume emit
> * When the `tokenURI` does not change, but the JSON does. <- I would also assume emit

The `MetadataUpdate` event MUST be emitted when the JSON metadata of a token is changed.

Not emitting `MetadataUpdate` event is RECOMMENDED when a token is minted.

Not emitting `MetadataUpdate` event is RECOMMENDED  when a token is burned.

Not emitting `MetadataUpdate` event is RECOMMENDED  when the tokenURI changes but the JSON metadata does not.

Reply:
Great, I think that clears up my concern. I think you can put that verbatim into the EIP.